### PR TITLE
provide custom marshaller for USER_INFO_1

### LINKED ISSUE #81705: Provide custom marshaller for `USER_INFO_1`
In PR https://github.com/dotnet/runtime/pull/81480, the following P/Invoke was failing to compile when targeting .NET 8.

https://github.com/dotnet/runtime/blob/167112574ebfcfcddfc23267e1ba93eebdcdeaf1/src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs#L121-L122

The failure is because in the current dotnet/runtime official build this P/Invoke is only compiled to target .NET 6 and so the `LibraryImport` source generated degrades to a `DllImport`. We should provide a custom marshaller for the struct below

https://github.com/dotnet/runtime/blob/167112574ebfcfcddfc23267e1ba93eebdcdeaf1/src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs#L127-L138

This scenario could be made to "just work" with the following issue addressed: https://github.com/dotnet/runtime/issues/81656

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #81705

### CODE REVIEW

File: src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs [hunk 1]
```diff
@@ -135,6 +141,73 @@ internal struct USER_INFO_1
             public string usri1_comment;
             public uint usri1_flags;
             public string usri1_script_path;
+
+#if NET7_0_OR_GREATER
+            [CustomMarshaller(typeof(USER_INFO_1), MarshalMode.Default, typeof(USER_INFO_1Marshaller))]
+            public static class USER_INFO_1Marshaller
```
Comment (xtqqczze) [inline src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs:147]:
```suggestion
            public static class Marshaller
```

File: src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs [hunk 2]
```diff
@@ -135,6 +141,73 @@ internal struct USER_INFO_1
...
+            {
+                public static USER_INFO_1Native ConvertToUnmanaged(USER_INFO_1 managed) => new(managed);
+                public static USER_INFO_1 ConvertToManaged(USER_INFO_1Native native) => native.ToManaged();
+                public static void Free(USER_INFO_1Native native) => native.FreeNative();
+
+                [StructLayout(LayoutKind.Explicit)]
```
Comment (xtqqczze) [inline src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs:153]:
```suggestion
                [StructLayout(LayoutKind.Sequential)]
```

`IntPtr.Size` is 4 on x86, so we can't use explicit layout,

We should ensure the struct layout is correct on 32 bit systems.

File: src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs (hunk 1 above)
Comment (pedrobsaila) [reply src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs:147]: Is there a constraint on Marshaller class naming to make it work ? or is it just a standard convention ?

File: src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs (hunk 2 above)
Comment (pedrobsaila) [reply src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs]: thanks for spotting this ;)

File: src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs (hunk 1 above)
Comment (xtqqczze) [reply src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs:147]: I just thought the name looked awkward and that `Marshaller` was the convention.

File: src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs [hunk 3]
```diff
@@ -118,12 +121,15 @@ private void CreateUser()
...
-        [DllImport("netapi32.dll", SetLastError = true)]
-        internal static extern uint NetUserAdd([MarshalAs(UnmanagedType.LPWStr)] string servername, uint level, ref USER_INFO_1 buf, out uint parm_err);
+        [LibraryImport("netapi32.dll", SetLastError = true)]
+        internal static partial uint NetUserAdd([MarshalAs(UnmanagedType.LPWStr)] string servername, uint level, ref USER_INFO_1 buf, out uint parm_err);
 
         [LibraryImport("netapi32.dll")]
         internal static partial uint NetUserDel([MarshalAs(UnmanagedType.LPWStr)] string servername, [MarshalAs(UnmanagedType.LPWStr)] string username);
 
+#if NET7_0_OR_GREATER
+        [NativeMarshalling(typeof(USER_INFO_1Marshaller))]
```
Comment (xtqqczze) [inline src/libraries/Common/tests/TestUtilities/System/WindowsIdentityFixture.cs:131]:
```suggestion
        [NativeMarshalling(typeof(USER_INFO_1.Marshaller))]
```

@pedrobsaila Could this avoid `CS0104` with the name `Marshaller`?

**Review by AaronRobinsonMSFT — APPROVED** [review]
Thank you!

**Review by jkoritzinsky — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
Failure is known https://github.com/dotnet/runtime/issues/81123.

**Pull request merged**